Mathlib/Algebra/QuadraticAlgebra/Basic.lean

```diff
@@ -0,0 +1,389 @@
+/-
+Copyright (c) 2025 Antoine Chambert-Loir. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambert-Loir
+-/
+
+import Mathlib.Algebra.QuadraticAlgebra.Defs
+import Mathlib.Algebra.Star.Unitary
+import Mathlib.Tactic.FieldSimp
+
+/-! # Quadratic algebras : involution and norm.
+
+Let `R` be a commutative ring. We define:
+
+* `QuadraticAlgebra.star` : the quadratic involution
+
+* `QuadraticAlgebra.norm` : the norm
+
+We prove :
+
+* `QuadraticAlgebra.isUnit_iff_norm_isUnit`:
+  `w : QuadraticAlgebra R a b` is a unit iff `w.norm` is a unit in `R`.
+
+* `QuadraticAlgebra.norm_mem_nonZero_divisors_iff`:
+  `w : QuadraticAlgebra R a b` isn't a zero divisor iff
+  `w.norm` isn't a zero divisor in `R`.
+
+* If `R` is a field, and `∀ r, r ^ 2 ≠ a + b * r`, then `QuadraticAlgebra R a b` is a field.
+`
+-/
+
+namespace QuadraticAlgebra
+
+variable {R : Type*} {a b : R}
+
+section omega
+
+section
+
+variable [Zero R] [One R]
+
+/-- The representative of the root in the quadratic algebra -/
+def ω : QuadraticAlgebra R a b :=
+  ⟨0, 1⟩
+
+@[simp]
+theorem re_omega : (ω : QuadraticAlgebra R a b).re = 0 :=
+  rfl
+
+@[simp]
+theorem im_omega : (ω : QuadraticAlgebra R a b).im = 1 :=
+  rfl
+
+end
+
+variable [CommSemiring R]
+
+theorem omega_mul_omega :
+    (ω : QuadraticAlgebra R a b) * (ω) = ⟨a, b⟩ := by
+  ext <;> simp
+
+theorem omega_mul_omega_eq :
+    (ω : QuadraticAlgebra R a b) * (ω) = a • 1 + b • (ω) := by
+  ext <;> simp
+
+@[simp]
+theorem omega_mul (x y : R) :
+    (ω : QuadraticAlgebra R a b) * ⟨x, y⟩ = ⟨a * y, x + b * y⟩ := by
+  ext <;> simp
+
+@[simp]
+theorem omega_mul_coe_mul (n x y : R) :
+    (ω : QuadraticAlgebra R a b) * (↑n) * ⟨x, y⟩ = ⟨a * n * y, n * x + n * b * y⟩ := by
+  ext <;> simp; ring
+
+theorem omega_prop : (ω) * (ω) =
+    (b : QuadraticAlgebra R a b) * (ω : QuadraticAlgebra R a b) + (a : QuadraticAlgebra R a b) := by
+  ext <;> simp
+
+theorem omega_prop' : (ω : QuadraticAlgebra R a b) * (ω) =
+    b • (ω) + a • 1 := by
+  simp [omega_prop, ← QuadraticAlgebra.coe_mul_eq_smul]
```

Comment:
Why are there so many variations on `ω * ω`, with very different names?

I think for `QuadraticAlgebra` you (i.e., Mathlib) should standardize on either `↑r * x` or `r • x` for `r : R` and `x : QuadraticAlgebra R a b`. Possibly there may be a special case for `r := a` or `r := b`, but in that case there should be additional `simp` lemmas for those.

Then there would only be 2 lemmas for `ω * ω`, the `mk` version and the `add` version (the latter using whichever standard we decide upon).